platforms/core-configuration/kotlin-dsl-plugins/src/main/kotlin/org/gradle/kotlin/dsl/plugins/embedded/EmbeddedKotlinPlugin.kt

```diff
@@ -63,14 +63,32 @@ abstract class EmbeddedKotlinPlugin @Inject internal constructor(
 
 fun Logger.warnOnDifferentKotlinVersion(kotlinVersion: String?) {
     if (kotlinVersion != embeddedKotlinVersion) {
-        warn(
-            """
-                WARNING: Unsupported Kotlin plugin version.
-                The `embedded-kotlin` and `kotlin-dsl` plugins rely on features of Kotlin `{}` that might work differently than in the requested version `{}`.
-            """.trimIndent(),
-            embeddedKotlinVersion,
-            kotlinVersion
-        )
+        val warning =
+            """|WARNING: Unsupported Kotlin plugin version.
+               |The `embedded-kotlin` and `kotlin-dsl` plugins rely on features of Kotlin `$embeddedKotlinVersion` that might work differently than in the requested version `$kotlinVersion`.
+               |
+               |Using the `kotlin-dsl` plugin together with a different Kotlin version (for example, by using the Kotlin Gradle plugin (`kotlin(jvm)`)) in the same module is not recommended.
+               |
+               |If you are writing a convention plugin (or a precompiled script plugin in general), it's recommended to use only the `kotlin-dsl` plugin and use the embedded Kotlin version. If you are writing a binary plugin, it's recommended that you do not use the `kotlin-dsl` plugin and use the Kotlin Gradle plugin (`kotlin(jvm)`).
+               |
+               |You can configure your module to use, for example, 2.1, by overriding `kotlin-dsl` behavior using the following snippet:
+               |
+               |    tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile> {
```

Comment:
❓  I think we should start by changing how the `kotlin-dsl` plugin configures Kotlin compiler options to make it easier to override its configuration. Build scripts should not have to configure tasks at all, ideally.

The `kotlin-dsl` plugin should instead configure Kotlin compiler options on the `kotlin` extension instead. Then build script can simply configure the extension to override the values.

This should be done as a separate PR before this one.